arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Racing.kt

```diff
@@ -0,0 +1,137 @@
+   *
+   * @param context The [CoroutineContext] to run the block in.
+   * @param block The coroutine block to race.
+   */
+  public suspend fun race(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> Result
+  )
+
+  /**
+   * Races a coroutine block against other blocks in the racing scope.
+   * If the block throws an exception, it will be propagated and cancel the entire racing scope.
+   * The block will be cancelled if another block completes first.
+   *
+   * @param context The [CoroutineContext] to run the block in.
+   * @param block The coroutine block to race.
+   */
+  public suspend fun raceOrThrow(
+    context: CoroutineContext = EmptyCoroutineContext,
+    block: suspend CoroutineScope.() -> Result
+  )
+}
+
+/**
+ * Creates a racing scope that allows multiple coroutine blocks to race against each other.
+ * The first block to complete successfully will provide the result, and all other blocks will be cancelled.
+ *
+ * This function provides structured concurrency guarantees - all racing blocks will be properly cancelled
+ * when the racing scope completes, either normally or exceptionally.
+ *
+ * Example:
+ * ```kotlin
+ * import arrow.fx.coroutines.racing
+ * import kotlinx.coroutines.delay
+ * import kotlin.time.Duration.Companion.milliseconds
+ *
+ * suspend fun main() {
+ *   val result = racing<String> {
+ *     // This block will be cancelled when the timeout completes
+ *     race {
+ *       delay(1000)
+ *       "Slow result"
+ *     }
+ *
+ *     // This block will complete first and provide the result
+ *     onTimeout(time = 100.milliseconds) {
+ *       "Timeout result"
+ *     }
+ *   }
+ *
+ *   println(result) // Prints: Timeout result
+ * }
+ * ```
+ *
+ * @param A The type of value that will be returned by the racing blocks.
+ * @param handleException A function to handle exceptions thrown by racing blocks. If null, exceptions will be handled
+ *                       by the current coroutine context's [CoroutineExceptionHandler] or printed to stderr.
+ * @param block The block of code to execute within the racing scope.
+ * @return The result of the first racing block to complete successfully.
+ */
+public suspend fun <A> racing(
+  handleException: (suspend (context: CoroutineContext, exception: Throwable) -> Unit)? = null,
+  block: suspend RacingScope<A>.() -> Unit,
+): A {
+  val exceptionHandler: suspend (context: CoroutineContext, exception: Throwable) -> Unit =
+    handleException
+      ?: currentCoroutineContext()[CoroutineExceptionHandler]?.let { { ctx, e -> it.handleException(ctx, e) } }
+      ?: { _, e -> e.printStackTrace() }
+  return coroutineScope {
+    val job = Job(currentCoroutineContext()[Job])
+    try {
+      select {
+        launch(start = CoroutineStart.UNDISPATCHED) {
+          val scope = SelectRacingScope(this@select, this@coroutineScope, job, exceptionHandler)
+          scope.block()
```

Comment:
It feels a little weird to potentially `suspend` within `select`

Reply:
Good point, probably better to remove this weird hack to support `suspend` there.